File: src/main/java/com/hazza/LeetCodeByJava/No39CombinationSum.java
package com.hazza.LeetCodeByJava;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * https://leetcode.com/problems/combination-sum/#/description
 */
public class No39CombinationSum {
    //solution 1
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<>();
        if (candidates == null || candidates.length == 0 || target < 0) {
            return res;
        }
        Arrays.sort(candidates);
        getCombination(candidates, 0, candidates.length, target,
                res, new ArrayList<>());

        return res;
    }

    private void getCombination(int[] candicates, int start, int end, int target,
                                List<List<Integer>> res, List<Integer> path) {
        if (target < 0) return;
        else if (target == 0) res.add(new ArrayList<>(path));    //be careful!!!
        else {
            for (int i = start; i < end; i++) {
                path.add(candicates[i]);
                getCombination(candicates, i, end, target - candicates[i], res, path);
                path.remove(path.size() - 1);
            }
        }
    }


}

File: src/main/python/leetcode_by_python/SegmentOne/No39_Combination_Sum.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Created by HazzaCheng on 2019/5/4 20:24
# https://leetcode.com/problems/combination-sum/
from typing import List


class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        self.dfs(candidates, target, 0, [], res)

        return res

    def dfs(self, candidates, target, idx, path, res):
        if target < 0:
            return
        if target == 0:
            res.append(path)
            return
        for i in range(idx, len(candidates)):
            self.dfs(candidates, target - candidates[i], i, path + [candidates[i]], res)


if __name__ == '__main__':
    solution = Solution()
    print(solution.combinationSum([2, 3, 6, 7], 7))
    assert sorted(solution.combinationSum([2, 3, 6, 7], 7)) == sorted([[7], [2, 2, 3]])
    assert sorted(solution.combinationSum([2, 3, 5], 8)) == sorted([[2, 2, 2, 2], [2, 3, 3], [3, 5]])